### ISSUE TITLE
ProjectAnalyser not finding all Metric Providers?

### ISSUE THREAD

**Issue by creat89** [crossminer/scava#188]
Hello,

I'm having a problem with ProjectAnalyser, which it doesn't seem to be finding all the metric providers. In fact, it is only finding 21 metrics:

```
org.eclipse.scava.factoid.bugs.channelusage.BugsChannelUsageFactoid
org.eclipse.scava.factoid.bugs.size.BugsChannelSizeFactoid
org.eclipse.scava.factoid.cocomo.CocomoFactoid
org.eclipse.scava.metricprovider.historic.bugs.comments.CommentsHistoricMetricProvider
org.eclipse.scava.metricprovider.historic.bugs.newbugs.NewBugsHistoricMetricProvider
org.eclipse.scava.metricprovider.historic.bugs.patches.PatchesHistoricMetricProvider
org.eclipse.scava.metricprovider.historic.commitsovertime.CommitsOverTimeHistoricMetricProvider
org.eclipse.scava.metricprovider.historic.configuration.puppet.designsmells.PuppetDesignHistoricMetricProvider
org.eclipse.scava.metricprovider.historic.newsgroups.articles.ArticlesHistoricMetricProvider
org.eclipse.scava.metricprovider.historic.numberofcommitters.model.CommittersOverTime
org.eclipse.scava.metricprovider.trans.bugs.comments.CommentsTransMetricProvider
org.eclipse.scava.metricprovider.trans.bugs.newbugs.NewBugsTransMetricProvider
org.eclipse.scava.metricprovider.trans.bugs.patches.PatchesTransMetricProvider
org.eclipse.scava.metricprovider.trans.commits.CommitsTransientMetricProvider
org.eclipse.scava.metricprovider.trans.committers.CommittersMetricProvider
org.eclipse.scava.metricprovider.trans.configuration.puppet.PuppetTransMetricProvider
org.eclipse.scava.metricprovider.trans.configuration.puppet.designsmells.PuppetDesignTransMetricProvider
org.eclipse.scava.metricprovider.trans.configuration.puppet.implementationsmells.PuppetImplementationTransMetricProvider
org.eclipse.scava.metricprovider.trans.dailycommits.DailyCommitsMetricProvider
org.eclipse.scava.metricprovider.trans.hourlycommits.HourlyCommitsMetricProvider
org.eclipse.scava.metricprovider.trans.newsgroups.articles.ArticlesTransMetricProvider
```

These are always returned, no more no less. The error could be similar to the one described in issue https://github.com/crossminer/scava/issues/116 and issue https://github.com/crossminer/scava/issues/134. However, there, or at least in https://github.com/crossminer/scava/issues/116, the problem occurred in the method `filterMetricProvider`, where for unknown reasons the platform couldn't match always the declared metrics and those found in the platform. However, the issue presented here occurs early in the code, more specifically in `platform.getMetricProviderManager().getMetricProviders()`. Here, unlike in issue https://github.com/crossminer/scava/issues/116, the platform is returning only 21 metrics. So, when it arrives to `filterMetricProvider`, there are only 21 metrics found in the platform.

At the end, this means that the platform cannot run all the declared metrics, because they are not, in theory, found in the platform. I have checked that the plug-ins are included, and yes. Everything seems to be declared as usual. I haven't either found any pattern. At the beginning, I though that they were metrics that do not depend on others (and maybe no other metrics depend on them), but the exception in this case is `org.eclipse.scava.metricprovider.trans.dailycommits.DailyCommitsMetricProvider` or `org.eclipse.scava.metricprovider.trans.hourlycommits.HourlyCommitsMetricProvider`. So, I'm lost in the sense that I don't know why this is happening.

 If I remember well, some versions ago, in the Mongo collection, they were stored all the metric providers. Now, I cannot see this any longer.

Anybody has observed this issue?

**Comment by ambpro** [comment]
> If I remember well, some versions ago, in the Mongo collection, they were stored all the metric providers. Now, I cannot see this any longer.

A few days ago, we did a refactoring to the way in which the platform loads the metric providers https://github.com/crossminer/scava/commit/24090c282c05749005a3574f1765f5f7300c1c17.
Before, the metric providers were loaded and stored in a mongodb collection each time the platform was launched. At runtime, the endpoint `/analysis/metricproviders` retrieved the list of the metric providers stored on the mongodb collection.
On the other hand, the endpoint `/raw/metrics` (relic from OSSMETER) access directly at runtime to the list of metrics via the extension point mechanism.
Due to the issue #143, we opted to reimplement the `/analysis/metricproviders` to reload the available metrics via the extension point mechanism on real time without restored the results on mongodb in order to fix the the inconsistent numbers of metrics.

**Comment by creat89** [comment]
OK, so why is it finding only 21 metrics? Is there anyway to know why these and not others? I'm sure there are more than 21 metrics with an extension point.

**Comment by tdegueul** [comment]
Strangely, I cannot reproduce this issue.

Both accessing the list of metrics using the REST API (ie. through `:8182/analysis/metricproviders` using [this code](https://github.com/crossminer/scava/blob/dev/metric-platform/tests/org.eclipse.scava.platform.tests/src/org/eclipse/scava/platform/tests/ApiHelper.java#L86-L103)) and using `platform.getMetricProviderManager().getMetricProviders()` return 315 metrics.

Can you please try using `platform.getAnalysisRepositoryManager().getMetricProviderService().getMetricProviders(platform)` (ie. as done [here](https://github.com/crossminer/scava/blob/dev/metric-platform/platform/org.eclipse.scava.platform.client.api/src/org/eclipse/scava/platform/client/api/AnalysisMetricProvidersResource.java#L28-L67)) and report on the results?

My guess (again :)) is that you're invoking the code from outside an Eclipse context. In this case the extension points aren't filled and most of the metrics are lost on the way.

**Comment by creat89** [comment]
I'll let you know about the results.

However, with respect to your last guess, is no. In this case I'm invoking the platform using the product file (although indicating that has to run as ApiServer and Worker). Because, one thing is to add elements to the database and another completely different to run the platform.

**Comment by creat89** [comment]
I have pulled the last version of Scava just to be sure that it is the last version. I'm using the run configuration defined by the Product file (and adding only -worker and -apiServer to the arguments only). And I haven't added nothing to the DB, in fact, I've dropped everything, so Scava populate everything in the initial state. When I call `localhost:8182/analysis/metricproviders` I still get 21 metrics: 

https://gist.github.com/creat89/46504b4f05af4edbceb2c24815859bae

@tdegueul, you told me last time, that if the platform is run outside an Eclipse context, then `Platform.getExtensionRegistry()` would be null. As I'm not running it outside an Eclipse context, in this case I'm getting this in `ExtensionPointHelper.java` (org.eclipse.scava.platform):

![image](https://user-images.githubusercontent.com/17121539/56909125-837f1980-6a9f-11e9-8d73-667d3fb3e9e3.png)

However, on the line that gets `configurationElements`, it only retrieves 21 configurations that are those related to the 21 metrics loaded.

I don't know if this is related, but most of the bugs that I've seen are related to the first execution. In other words, I run always Scava as if wasn't run before. In other bugs, such as https://github.com/crossminer/scava/issues/134, we needed to run a second time Scava in order to detect new metrics. Could this be somehow related?
